src/main.rs

```diff
@@ -1,35 +1,690 @@
 //! ./src/main.rs
```

Comment:
A note on traits (used in comments below):

Traits are rusts version of interfaces. An interface is a set of functions which you need to satisfy in order to "count as" that interface. Let me clarify that with an example. Say we have an interface for communication:
```rust
trait Communicator {
    fn send(self, String);
    fn check_for_received_message(self) -> Option<String>;
}
```
This means that we have a set of behaviors/functions (sending and receiving), which, if satisfied, can make you count as a communicator. Then structs can implement this trait if they can behave as communicators
```rust
// struct representing some wire
struct TelegramWire { ... }

impl Communicator for TelegramWire {
    fn send(self, message: String) {
        self.telegram_tapper_thing.tap_out(message);
    }
    fn check_for_received_message(self) -> Option<String> {
        self.messages_from_earlier.next()
    }
}


// struct representing some wire
struct YellingPerson { ... }

impl Communicator for YellingPerson {
    fn send(self, message: String) {
        self.yell(message);
    }
    fn check_for_received_message(self) -> Option<String> {
        self.messages_i_heard.next()
    }
}
```
The trait/interface defines what needs to be implemented to be a communicator, and any struct with a way to implement those can then count as a communicator. This lets us write reusable code so that a function can accept any type that is able to meet certain behaviors.
```rust
fn notify_recipient_of_gift<C: Communicator, T>(communicator: C, gw: GiftWrap<T>) {
    communicator.send(gw.read_tag);
}
```
Here we have a type constraint `C: Communicator` which means "Any type C, as long as it can act like a communicator (ie send messages)". The `T` generic does not have any type constraint because we dont care what is in the giftwrap, becauase regardless the gift wrap has a tag.